### PR TITLE
[improve][client] PIP-218: Consumer batchReceive() single partition every receive

### PR DESCRIPTION
**Pull request by congbobo184**
### Motivation
https://github.com/apache/pulsar/issues/18182

### Modifications

[](https://github.com/apache/pulsar/issues/18182)

### Verifying this change
add test for batchReceive single topic messages
add test for batchReceive multi-topic messages

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [x] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [x] `doc-required` 
- [ ] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository
PR in forked repository: 

- https://github.com/congbobo184/pulsar/pull/5

### CODE REVIEW

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/BatchReceivePolicy.java [hunk 1]
```diff
@@ -114,6 +130,7 @@ public static class Builder {
         private int maxNumBytes;
         private int timeout;
         private TimeUnit timeoutUnit;
+        private boolean messagesFromMultiTopicsEnabled;
```
Comment (WJL3333) [inline pulsar-client-api/src/main/java/org/apache/pulsar/client/api/BatchReceivePolicy.java:133]: need default value here

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/BatchReceivePolicy.java [hunk 2]
```diff
@@ -83,6 +88,13 @@ private BatchReceivePolicy(int maxNumMessages, int maxNumBytes, int timeout, Tim
     private final int timeout;
     private final TimeUnit timeoutUnit;
 
+
+    /**
+     * If it is false, one time `batachReceive()` only can receive the single topic messages,
```
Comment (BewareMyPower) [inline pulsar-client-api/src/main/java/org/apache/pulsar/client/api/BatchReceivePolicy.java:93]:
```suggestion
     * If it is false, one time `batchReceive()` only can receive the single topic messages,
```

Typo.

File: pulsar-broker/src/test/java/org/apache/pulsar/client/api/ConsumerBatchReceiveTest.java [hunk 3]
```diff
@@ -648,5 +647,80 @@ private void batchReceiveAndRedelivery(Consumer<String> consumer, int expected)
...
+                    if (topicName != null) {
+                        // receive the different topic messages in one batch receive
+                        if (!topicName.equals(message.getTopicName())) {
+                            return;
+                        }
+                    }
+                    topicName = message.getTopicName();
+                }
+            }
+        }
```
Comment (liangyepianzhou) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/api/ConsumerBatchReceiveTest.java:722]: Good work! Leave a little suggestion.
Maybe we should add an `Assert.fail()` here.
If the multiTopicBatchReceiveConsumer can not receive the multiple topics messages.

**Review by liangyepianzhou — APPROVED** [review]

**Review by WJL3333 — APPROVED** [review]

**Review by poorbarcode — APPROVED** [review]

**Comment by momo-jun** [comment]
Hi @congbobo184, Did you have any plans for updating the docs?

**Pull request merged**